certora/specs/EnumerableMap.spec

```
@@ -0,0 +1,308 @@
+
+    uint256 lengthBefore        = length();
+    bool    containsBefore      = contains(key);
+    bool    containsOtherBefore = contains(otherKey);
+    bytes32 otherValueBefore    = tryGet_value(otherKey);
+
+    bool added = set@withrevert(key, value);
+    bool success = !lastReverted;
+
+    // liveness & immediate effect
+    assert success
+        && contains(key)
+        && get(key) == value;
+
+    // return value: added iff not contained
+    assert added <=> !containsBefore;
+
+    // effect: length increass iff added
+    assert length() == lengthBefore + to_uint256(added ? 1 : 0);
+
+    // effect: add at the end
+    assert added => (
+        at_key(lengthBefore) == key &&
+        at_value(lengthBefore) == value
+    );
+
+    // side effect: other keys are not affected
+    assert containsOtherBefore != contains(otherKey) => (added && key == otherKey);
+    assert otherValueBefore != tryGet_value(otherKey) => key == otherKey;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: remove key from EnumerableMap                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule remove(bytes32 key, bytes32 otherKey) {
+    requireInvariant consistencyKey(key);
+    requireInvariant consistencyKey(otherKey);
+
+    uint256 lengthBefore        = length();
+    bool    containsBefore      = contains(key);
+    bool    containsOtherBefore = contains(otherKey);
+    bytes32 otherValueBefore    = tryGet_value(otherKey);
+
+    bool removed = remove@withrevert(key);
+    bool success = !lastReverted;
+
+    // liveness & immediate effect
+    assert success && !contains(key);
+
+    // return value: removed iff contained
+    assert removed <=> containsBefore;
+
+    // effect: length increass iff removed
+    assert length() == lengthBefore - to_uint256(removed ? 1 : 0);
+
+    // side effect: other keys are not affected
+    assert containsOtherBefore != contains(otherKey) => (removed && key == otherKey);
+    assert otherValueBefore != tryGet_value(otherKey) => key == otherKey;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: when adding a new key, the other keys remain in set, at the same index.                                      │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule atSet(bytes32 key, bytes32 value, uint256 index) {
+    require sanity();
+
+    bytes32 atKeyBefore = at_key(index);
+    bytes32 atValueBefore = at_value(index);
+
+    set(key, value);
+
+    bytes32 atKeyAfter = at_key@withrevert(index);
+    assert !lastReverted;
+
+    bytes32 atValueAfter = at_value@withrevert(index);
+    assert !lastReverted;
```

I think we can remove the `@withrevert` here because it's proven that they only revert if there's no value

```suggestion
    bytes32 atKeyAfter = at_key(index);
    bytes32 atValueAfter = at_value(index);
```

---

Its only proven that it revert if the index is out of bound. And true, since we are doing set, the length increasses and so the function should not revert.

Strictly speaking this is already proven somewhere else. I still think its nice to include it here, just so that this rule fully verify that if a value was accessible before, it remains accessible at the same place. (without falling back to another rule)